pkg/sql/sqlinstance/instanceprovider/instanceprovider.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// Package instanceprovider provides an implementation
+// of the sqlinstance.provider interface
+package instanceprovider
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
+
+type writer interface {
+	CreateInstance(context.Context, sqlliveness.SessionID, string) (sqlinstance.SQLInstance, error)
+	ReleaseInstanceID(context.Context, base.SQLInstanceID) error
+}
+type storage interface {
+	sqlinstance.AddressResolver
+	writer
+	Start()
+}
+
+type provider struct {
```

Comment:
It'd be good to have a block of commentary here explaining what a "provider" is and how to use one.